cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -672,11 +678,64 @@ struct Utf8TitleTransform : public FunctionalCaseMappingTransform {
 template <typename Type>
 using Utf8Title = StringTransformExec<Type, Utf8TitleTransform>;
 
+struct Utf8NormalizeTransform : public FunctionalCaseMappingTransform {
+  using State = OptionsWrapper<Utf8NormalizeOptions>;
+
+  const Utf8NormalizeOptions* options;
+
+  explicit Utf8NormalizeTransform(const Utf8NormalizeOptions& options)
+      : options{&options} {}
+
+  int64_t MaxCodeunits(const uint8_t* input, int64_t ninputs,
+                       int64_t input_ncodeunits) override {
+    const auto option = GenerateUtf8NormalizeOption(options->method);
+    const auto n_chars =
+        utf8proc_decompose_custom(input, input_ncodeunits, NULL, 0, option, NULL, NULL);
+
+    // convert to byte length
+    return n_chars * 4;
+  }
+
+  int64_t Transform(const uint8_t* input, int64_t input_string_ncodeunits,
+                    uint8_t* output, int64_t output_string_ncodeunits) {
+    const auto option = GenerateUtf8NormalizeOption(options->method);
+    const auto n_chars = utf8proc_decompose_custom(
+        input, input_string_ncodeunits, reinterpret_cast<utf8proc_int32_t*>(output),
+        output_string_ncodeunits, option, NULL, NULL);
+    if (n_chars < 0) return output_string_ncodeunits;
+
+    const auto n_bytes =
+        utf8proc_reencode(reinterpret_cast<utf8proc_int32_t*>(output), n_chars, option);
+    return n_bytes;
+  }
```

Comment:
Several comments:
1. Why not use [`utf8proc_NFC()` and friends](https://juliastrings.github.io/utf8proc/doc/utf8proc_8h.html#a11717c1d4b4725be52102a1cc5cd1d3c)? These expect a null-terminated string so these would require copying the input string and null-terminating it. If this approach is used, then `Utf8NormalizeOptions` do not need to be resolved in constructor but here in the kernel.
2. [`utf8proc_reencode()`](https://juliastrings.github.io/utf8proc/doc/utf8proc_8h.html#a5e7debdee382d00b237d75f0932a3e73) has a warning message regarding its size and IIUC it creates a null-terminating string which you are writing into the output. AFAIK, Arrow stores strings contiguously without null-terminations.

Reply:
As for #1, avoiding a temporary copy (and allocation!) for each input string sounds beneficial.

As for #2, you are right. We can use our own UTF8 encoding instead (see `UTF8Encode` in `arrow/util/utf8.h`)